Add a mechanism for swapping builtin implementations for unison code

dolio opened the pull request:
This extends the interpreter with some systems for swapping out unison code with directly implemented builtins. It's possible to switch a unison data type with a builtin one, and to switch out unison implementations of functions with builtins by hash.

In this PR, the main target has been the `data.Map` type in base. It gets switched out for a wrapped Haskell `Map` from containers, and a few functions on it have been switched out for their Haskell equivalents. I also replaced a couple list functions, since they don't require a new type.

Benchmark timings:

```
fib1
238.779µs -> 226.183µs                          
                                  
fib2
3.869228ms -> 3.816003ms                          
                                  
fib3
4.301055ms -> 4.226743ms                          
                                  
Decode Nat
226ns -> 212ns                          
                                  
Generate 100 random numbers
231.076µs -> 236.088µs                          
                                  
List.foldLeft
2.25254ms -> 2.377862ms                         
                                  
Count to 1 million
51.69272ms -> 51.694017ms                         
                                  
Json parsing (per document)
371.181µs -> 363.121µs                          
                                  
Count to N (per element)
123ns -> 121ns                          
                                  
Count to 1000
121.204µs -> 120.034µs                          
                                  
Mutate a Ref 1000 times
183.651µs -> 189.658µs                          
                                  
CAS an IO.ref 1000 times
248.142µs -> 223.604µs                          
                                  
List.range (per element)
313ns -> 112ns                          
                                  
List.range 0 1000
412.416µs -> 154.091µs                          
                                  
Set.fromList (range 0 1000)
1.98408ms -> 488.404µs                          
                                  
Map.fromList (range 0 1000)
1.306564ms -> 51.956µs                            
                                  
NatMap.fromList (range 0 1000)
4.846166ms -> 4.674058ms                          
                                  
Map.lookup (1k element map)
2.053µs -> 260ns                            
                                  
Map.insert (1k element map)
6.099µs -> 432ns                            
                                  
List.at (1k element list)
217ns -> 229ns                          
                                  
Text.split /
62.525µs -> 57.521µs                          
```

dolio committed:
Add unison Map definition to Decls

This lets us look up the hash/constructors programmatically.

dolio committed:
Add a ref and BuiltinForeign for wrapped Haskell maps

dolio committed:
Allow the foreign maps to act as data types in matches

dolio committed:
Move universal comparisons to Stack, add some map stuff

Moving the comparison/equality to the Stack module allows the universal
operations to be used as the Eq/Ord instance for `Val`, which allows
those to work properly in Maps.

Map handling has been added to the universal eq/ord, to emulate how they
would be compared as unison values, rather than Map equality.

Functions for mediating between Haskell Map and the unison datatype
representation have been added, for use in decompiling and serializing.

dolio committed:
Decompile foreign maps as the corresponding Unison representation

dolio committed:
Implement ANF transformations for replacing constructors/functions

dolio committed:
Add foreign functions for interacting with maps

Pseudo constructors and some replacements for complex functions.

dolio committed:
Implement the replacement optimization in the interpreter

This replaces the operations _after_ recording the code, so it will
be invisible at the interchange level.

dolio committed:
Swap out map representations when reifying/reflecting values

dolio committed:
Add some additional function replacements

dolio commented:
I think all my usual tests have passed. Also `+optchecks` passes. Hopefully Cody or Stew can report on the cloud tests (and how much good this patch does there).

dolio commented:
I'm uncertain what the macos failure means. Is that something I did?

pchiusano reviewed (approved):
Awesome!

This is a really nice speedup for maps, and also the general capability is super useful as well.

ceedubs reviewed (changes requested):
Hold off on merging this. The Unison Cloud integration tests are showing an issue. I'll be in touch with Dan.

dolio committed:
Use incremental replacement when reifying data

The scheme before was trying to decode as normal then replace,
but that would result in only replacing leaves. It doesn't seem
feasible to do a top-down approach, so instead each value is
replaced incrementally from the bottom up.

dolio commented:
@ceedubs I tweaked the way value loading works. I think it wasn't correct for non-trivial maps before.

Let me know if there are still test failures.

ceedubs commented:
@dolio sure I'll kick off some more tests.

Value loading nontrivial maps is something that happens quite a bit in Unison Cloud. It might be worth adding a "golden file" style test with a serialized map that gets deserialized and loaded.

ceedubs reviewed (approved):
I still think that some test coverage would be great, but I can confirm that the latest commit fixed the issue that I saw.

ChrisPenner commented on unison-runtime/src/Unison/Runtime/Foreign.hs:
```diff
@@ -358,6 +359,15 @@ instance BuiltinForeign CharPattern where
   foreignName = Tagged "CharPattern"
   foreignRef = Tagged Ty.charClassRef
 
+-- Note: this doesn't do any recursive conversion of keys/values,
+-- so any use of it needs to be an exact match for the map that was
+-- originally placed in the box. The current intention is for use
+-- at `Map Val Val`, but `Val` is in a module further along the
+-- dependency graph.
```
Would it maybe be better to just define the `Map Val Val` instance in the module with `Val` then? It's nice to avoid footguns when we can :)

dolio replied:
Hmm, maybe.

ChrisPenner commented on unison-runtime/src/Unison/Runtime/Machine.hs:
```diff
@@ -877,6 +879,13 @@ dataBranch mrf stk (Test1 u cu df) = \case
     | maskTags t == u -> (cu,) <$> dumpSeg stk seg S
     | otherwise -> pure (df, stk)
+  Foreign f
+    | Just m <- maybeUnwrapForeign Rf.hmapRef f -> case m of
```
Seems like it could be beneficial to add pattern synonyms for each of the types we rewrite so we can benefit from the pattern-match-coverage checker wherever they're used. That'd be my preference at least :)

dolio replied:
I could make some patterns. I'm unsure what you mean by the coverage checker, though. No set of patterns is going to cover all the foreign values, so there need to be default cases, which means there isn't great coverage checking.

Did you have something more elaborate in mind?

ChrisPenner reviewed (approved):
Looks good! Good work on this!

For types that we rewrite like this, what happens if the hash of the unison type changes? I assume it'll just fall back to the unison implementations since the hashes won't match?
If that's the case, is there anywhere we can add checks to assert the optimizations are kicking in on current base? I'm guessing not, but if we plan to do this with more types it might be worthwhile to make sure we're not accidentally introducing changes which break optimizations. Guess it's fine without for now though :)

pchiusano commented:
💯 on adding a test here that would have caught the round trip error that was (incidentally) caught by cloud integration tests.

dolio commented:
The reason I didn't add a test is that you need the replaced `Map` functions to see the error.

Maybe `lookup` isn't too bad, though. Unless someone can think of a better way by tomorrow.

dolio committed:
Add a roundtrip test for replacements

dolio committed:
Move builtin decompile case below general reference case

This allows backmapping special builtins to the functions they replace

dolio committed:
Handle partial applications of replaced functions better

Record backmappings from replacements to their original functions.

During reify/reflect of partial applications, use the replacement
mappings to mediate between reflection values that refer to the
_original_ function, not the replacement. Reflection uses the
backmap and reification uses the forward map.

Decompiling also uses the backmap to avoid displaying the hidden
builtins, in favor of the unison implementation.

dolio committed:
Add a test for partial applications of replaced functions

Checks that a watch with a partial application displays code, not
the replacement builtin.

dolio committed:
Switch to update Map.insert hash

dolio commented:
@ceedubs Probably could use one more trip through the cloud tests before merging.

ceedubs reviewed (approved):
Cloud tests are all passing.

I did hit one issue the first time that I ran them, but I think that it was just a timeout while my computer was still working on a haskell build. They ran successfully 4 or 5 times after that, so I think it was unrelated.
